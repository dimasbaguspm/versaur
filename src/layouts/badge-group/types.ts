import type { HTMLAttributes, ReactNode } from 'react'
import type { VariantProps } from '@/utils/variants'
import type { badgeGroupVariants } from './helpers'

/**
 * Props for the BadgeGroup component
 */
export interface BadgeGroupProps
  extends HTMLAttributes<HTMLDivElement>,
    VariantProps<typeof badgeGroupVariants> {
  /**
   * Children (should be Badge components)
   */
  children: ReactNode

  /**
   * Alignment of badges within the group
   * @default 'start'
   */
  alignment?: 'start' | 'center' | 'end' | 'between' | 'around' | 'evenly'

  /**
   * Whether badges should be fluid (full width)
   * @default false
   */
  fluid?: boolean

  /**
   * Orientation of the badge group
   * @default 'horizontal'
   */
  orientation?: 'horizontal' | 'vertical'

  /**
   * Size of the gap between badges
   * @default 'md'
   */
  gap?: 'xs' | 'sm' | 'md' | 'lg' | 'xl'

  /**
   * Whether the badge group has bottom margin
   * @default false
   */
  hasMargin?: boolean
}
