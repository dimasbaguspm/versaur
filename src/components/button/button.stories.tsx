import type { Meta, StoryObj } from '@storybook/react'
import { Button } from './button'

const meta: Meta<typeof Button> = {
  title: 'Components/Button',
  component: Button,
  parameters: {
    layout: 'centered',
    docs: {
      description: {
        component: `
The Button component is a versatile, accessible button with support for semantic variants, outline styles, ghost styles, and multiple sizes. Built with Tailwind CSS and following Versaur's warm, earthy design system.

## Features
- **Semantic Variants**: Primary, secondary, tertiary, success, info, warning, danger
- **Style Variants**: Filled, outline, ghost
- **Sizes**: Small, medium, large
- **Accessibility**: Full keyboard navigation, ARIA attributes, focus management
- **Responsive**: Mobile-first responsive design

## Color Harmony
Uses Versaur's harmonious color palette with warm, earthy tones that create natural visual flow and excellent accessibility.
        `,
      },
    },
  },
  argTypes: {
    variant: {
      control: { type: 'select' },
      options: [
        'primary',
        'primary-outline',
        'primary-ghost',
        'secondary',
        'secondary-outline',
        'secondary-ghost',
        'tertiary',
        'tertiary-outline',
        'tertiary-ghost',
        'success',
        'success-outline',
        'success-ghost',
        'info',
        'info-outline',
        'info-ghost',
        'warning',
        'warning-outline',
        'warning-ghost',
        'danger',
        'danger-outline',
        'danger-ghost',
        'outline',
        'ghost',
        'destructive',
      ],
      description: 'Visual style variant of the button',
    },
    size: {
      control: { type: 'select' },
      options: ['sm', 'md', 'lg'],
      description: 'Size of the button',
    },
    disabled: {
      control: { type: 'boolean' },
      description: 'Whether the button is disabled',
    },
    children: {
      control: { type: 'text' },
      description: 'Button content',
    },
  },
}

export default meta
type Story = StoryObj<typeof Button>

// =============================================================================
// Color Stories - Semantic color system
// =============================================================================

/**
 * Complete color matrix showing all semantic colors in all style variants.
 * Demonstrates the comprehensive color system and variant combinations.
 */
export const ColorMatrix: Story = {
  render: () => (
    <div className='space-y-6'>
      <div>
        <h3 className='text-lg font-semibold mb-3'>Filled Variants</h3>
        <div className='flex flex-wrap gap-3'>
          <Button variant='primary'>Primary</Button>
          <Button variant='secondary'>Secondary</Button>
          <Button variant='tertiary'>Tertiary</Button>
          <Button variant='success'>Success</Button>
          <Button variant='info'>Info</Button>
          <Button variant='warning'>Warning</Button>
          <Button variant='danger'>Danger</Button>
        </div>
      </div>

      <div>
        <h3 className='text-lg font-semibold mb-3'>Outline Variants</h3>
        <div className='flex flex-wrap gap-3'>
          <Button variant='primary-outline'>Primary</Button>
          <Button variant='secondary-outline'>Secondary</Button>
          <Button variant='tertiary-outline'>Tertiary</Button>
          <Button variant='success-outline'>Success</Button>
          <Button variant='info-outline'>Info</Button>
          <Button variant='warning-outline'>Warning</Button>
          <Button variant='danger-outline'>Danger</Button>
        </div>
      </div>

      <div>
        <h3 className='text-lg font-semibold mb-3'>Ghost Variants</h3>
        <div className='flex flex-wrap gap-3'>
          <Button variant='primary-ghost'>Primary</Button>
          <Button variant='secondary-ghost'>Secondary</Button>
          <Button variant='tertiary-ghost'>Tertiary</Button>
          <Button variant='success-ghost'>Success</Button>
          <Button variant='info-ghost'>Info</Button>
          <Button variant='warning-ghost'>Warning</Button>
          <Button variant='danger-ghost'>Danger</Button>
        </div>
      </div>
    </div>
  ),
}

/**
 * All semantic colors in their filled variant.
 * Demonstrates the warm, earthy color palette and harmonious relationships.
 */

export const SemanticColors: Story = {
  render: () => (
    <div className='flex flex-wrap gap-4'>
      <Button variant='primary'>Primary</Button>
      <Button variant='secondary'>Secondary</Button>
      <Button variant='tertiary'>Tertiary</Button>
      <Button variant='success'>Success</Button>
      <Button variant='info'>Info</Button>
      <Button variant='warning'>Warning</Button>
      <Button variant='danger'>Danger</Button>
    </div>
  ),
}

/**
 * Outline variants provide a lighter visual weight while maintaining semantic meaning.
 * Perfect for secondary actions or when you need multiple buttons in close proximity.
 */
export const OutlineVariants: Story = {
  render: () => (
    <div className='flex flex-wrap gap-4'>
      <Button variant='primary-outline'>Primary Outline</Button>
      <Button variant='secondary-outline'>Secondary Outline</Button>
      <Button variant='tertiary-outline'>Tertiary Outline</Button>
      <Button variant='success-outline'>Success Outline</Button>
      <Button variant='info-outline'>Info Outline</Button>
      <Button variant='warning-outline'>Warning Outline</Button>
      <Button variant='danger-outline'>Danger Outline</Button>
    </div>
  ),
}

/**
 * Ghost variants are the most subtle, perfect for navigation or when you need
 * a button that doesn't visually compete with other content.
 */
export const GhostVariants: Story = {
  render: () => (
    <div className='flex flex-wrap gap-4'>
      <Button variant='primary-ghost'>Primary Ghost</Button>
      <Button variant='secondary-ghost'>Secondary Ghost</Button>
      <Button variant='tertiary-ghost'>Tertiary Ghost</Button>
      <Button variant='success-ghost'>Success Ghost</Button>
      <Button variant='info-ghost'>Info Ghost</Button>
      <Button variant='warning-ghost'>Warning Ghost</Button>
      <Button variant='danger-ghost'>Danger Ghost</Button>
    </div>
  ),
}

/**
 * Comparison of all three style variants for the primary semantic color.
 * Shows how the same semantic meaning can be expressed with different visual weights.
 */
export const StyleComparison: Story = {
  render: () => (
    <div className='flex flex-wrap gap-4'>
      <Button variant='primary'>Primary Filled</Button>
      <Button variant='primary-outline'>Primary Outline</Button>
      <Button variant='primary-ghost'>Primary Ghost</Button>
    </div>
  ),
}

// =============================================================================
// Size Stories - Different button sizes
// =============================================================================

/**
 * Size comparison showing all three sizes side by side.
 * Demonstrates the proportional scaling of button sizes.
 */
export const SizeComparison: Story = {
  render: () => (
    <div className='flex flex-wrap items-center gap-4'>
      <Button size='sm'>Small</Button>
      <Button size='md'>Medium</Button>
      <Button size='lg'>Large</Button>
    </div>
  ),
}

// =============================================================================
// State Stories - Interactive states and conditions
// =============================================================================

/**
 * Comparison of enabled vs disabled states across different variants.
 * Shows how disabled state affects visual appearance while maintaining accessibility.
 */
export const DisabledComparison: Story = {
  render: () => (
    <div className='space-y-4'>
      <div className='flex flex-wrap gap-4'>
        <Button variant='primary'>Enabled</Button>
        <Button variant='primary' disabled>
          Disabled
        </Button>
      </div>
      <div className='flex flex-wrap gap-4'>
        <Button variant='primary-outline'>Enabled Outline</Button>
        <Button variant='primary-outline' disabled>
          Disabled Outline
        </Button>
      </div>
      <div className='flex flex-wrap gap-4'>
        <Button variant='primary-ghost'>Enabled Ghost</Button>
        <Button variant='primary-ghost' disabled>
          Disabled Ghost
        </Button>
      </div>
    </div>
  ),
}

/**
 * Loading state example showing how buttons can indicate ongoing processes.
 * While not built into the component, shows common usage patterns.
 */
export const LoadingState: Story = {
  args: {
    children: 'Loading...',
    disabled: true,
  },
}
