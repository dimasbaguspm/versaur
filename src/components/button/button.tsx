import React from 'react'
import { cn } from '@/utils/cn'
import { cva } from '@/utils/variants'
import type { ButtonProps } from './types'

const buttonVariants = cva(
  'inline-flex items-center justify-center gap-2 rounded-md font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none',
  {
    variants: {
      variant: {
        // Core variants using Versaur color system
        primary:
          'bg-coral text-white hover:bg-coral/90 focus-visible:ring-coral-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md',
        secondary:
          'bg-sage text-white hover:bg-sage/90 focus-visible:ring-sage-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md',
        tertiary:
          'bg-mist text-white hover:bg-mist/90 focus-visible:ring-mist-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md',
        ghost:
          'bg-white text-foreground hover:bg-slate-soft focus-visible:ring-slate-soft focus-visible:ring-offset-white',
        neutral:
          'bg-neutral text-foreground border border-border hover:bg-neutral/80 focus-visible:ring-foreground-soft focus-visible:ring-offset-white shadow-sm',

        // Outline variants
        'primary-outline':
          'border border-coral text-coral bg-white hover:bg-coral hover:text-white focus-visible:ring-coral-soft focus-visible:ring-offset-white transition-all',
        'secondary-outline':
          'border border-sage text-sage bg-white hover:bg-sage hover:text-white focus-visible:ring-sage-soft focus-visible:ring-offset-white transition-all',
        'tertiary-outline':
          'border border-mist text-mist bg-white hover:bg-mist hover:text-white focus-visible:ring-mist-soft focus-visible:ring-offset-white transition-all',
        'ghost-outline':
          'border border-slate text-slate bg-white hover:bg-slate hover:text-white focus-visible:ring-slate-soft focus-visible:ring-offset-white transition-all',
        'neutral-outline':
          'border border-neutral text-foreground bg-white hover:bg-neutral hover:text-foreground focus-visible:ring-foreground-soft focus-visible:ring-offset-white transition-all',

        // Ghost variants (subtle)
        'primary-ghost':
          'text-coral bg-white hover:bg-coral-soft focus-visible:ring-coral-soft focus-visible:ring-offset-white',
        'secondary-ghost':
          'text-sage bg-white hover:bg-sage-soft focus-visible:ring-sage-soft focus-visible:ring-offset-white',
        'tertiary-ghost':
          'text-mist bg-white hover:bg-mist-soft focus-visible:ring-mist-soft focus-visible:ring-offset-white',
        'neutral-ghost':
          'text-foreground bg-white hover:bg-neutral/70 focus-visible:ring-foreground-soft focus-visible:ring-offset-white',

        // Semantic variants
        success:
          'bg-success text-white hover:bg-success/90 focus-visible:ring-success-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md',
        'success-outline':
          'border border-success text-success bg-white hover:bg-success hover:text-white focus-visible:ring-success-soft focus-visible:ring-offset-white transition-all',
        'success-ghost':
          'text-success bg-white hover:bg-success-soft focus-visible:ring-success-soft focus-visible:ring-offset-white',

        info: 'bg-info text-white hover:bg-info/90 focus-visible:ring-info-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md',
        'info-outline':
          'border border-info text-info bg-white hover:bg-info hover:text-white focus-visible:ring-info-soft focus-visible:ring-offset-white transition-all',
        'info-ghost':
          'text-info bg-white hover:bg-info-soft focus-visible:ring-info-soft focus-visible:ring-offset-white',

        warning:
          'bg-warning text-white hover:bg-warning/90 focus-visible:ring-warning-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md',
        'warning-outline':
          'border border-warning text-warning bg-white hover:bg-warning hover:text-white focus-visible:ring-warning-soft focus-visible:ring-offset-white transition-all',
        'warning-ghost':
          'text-warning bg-white hover:bg-warning-soft focus-visible:ring-warning-soft focus-visible:ring-offset-white',

        danger:
          'bg-danger text-white hover:bg-danger/90 focus-visible:ring-danger-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md',
        'danger-outline':
          'border border-danger text-danger bg-white hover:bg-danger hover:text-white focus-visible:ring-danger-soft focus-visible:ring-offset-white transition-all',
        'danger-ghost':
          'text-danger bg-white hover:bg-danger-soft focus-visible:ring-danger-soft focus-visible:ring-offset-white',

        // Utility variants
        outline:
          'border border-border text-foreground bg-white hover:bg-accent-soft focus-visible:ring-accent-soft focus-visible:ring-offset-white transition-all',
        destructive:
          'bg-danger text-white hover:bg-danger/90 focus-visible:ring-danger-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md',
      },
      size: {
        sm: 'h-8 px-3 text-sm min-w-[2.25rem]',
        md: 'h-10 px-4 text-sm min-w-[2.5rem]',
        lg: 'h-12 px-8 text-lg min-w-[2.75rem]',
      },
    },
    defaultVariants: {
      variant: 'primary',
      size: 'md',
    },
  }
)

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  function Button(
    {
      className,
      variant = 'primary',
      size = 'md',
      disabled = false,
      type = 'button',
      children,
      ...props
    },
    ref
  ) {
    return (
      <button
        ref={ref}
        type={type}
        className={cn(buttonVariants({ variant, size }), className)}
        disabled={disabled}
        aria-disabled={disabled}
        inert={disabled ? true : undefined}
        {...props}
      >
        {children}
      </button>
    )
  }
)
