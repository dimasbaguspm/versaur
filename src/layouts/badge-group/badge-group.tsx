import { cn } from '@/utils/cn'
import { badgeGroupVariants } from './helpers'
import type { BadgeGroupProps } from './types'
import { forwardRef } from 'react'

/**
 * BadgeGroup provides layout management for badges with control over
 * positioning, alignment, and fluid behavior
 *
 * @example
 * <BadgeGroup alignment="center" gap="sm">
 *   <Badge variant="primary">Save</Badge>
 *   <Badge variant="ghost">Cancel</Badge>
 * </BadgeGroup>
 *
 * @example
 * // Vertical badge group
 * <BadgeGroup orientation="vertical" fluid>
 *   <Badge variant="primary">Save</Badge>
 *   <Badge variant="secondary">Save as Draft</Badge>
 *   <Badge variant="ghost">Cancel</Badge>
 * </BadgeGroup>
 *
 * @example
 * // Fluid badges with space between
 * <BadgeGroup alignment="between" fluid>
 *   <Badge variant="ghost">Back</Badge>
 *   <Badge variant="primary">Next</Badge>
 * </BadgeGroup>
 */
export const BadgeGroup = forwardRef<HTMLDivElement, BadgeGroupProps>(
  function BadgeGroup(
    {
      className,
      children,
      orientation = 'horizontal',
      alignment = 'start',
      gap = 'md',
      fluid = false,
      ...props
    },
    ref
  ) {
    return (
      <div
        ref={ref}
        className={cn(
          badgeGroupVariants({
            orientation,
            alignment,
            gap,
            fluid,
          }),
          className
        )}
        role='group'
        {...props}
      >
        {children}
      </div>
    )
  }
)
