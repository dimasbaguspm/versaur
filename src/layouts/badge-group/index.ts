export { BadgeGroup } from './badge-group'
export type { BadgeGroupProps } from './types'
