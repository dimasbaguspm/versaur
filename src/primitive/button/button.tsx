import React from 'react'
import { cn } from '@/utils/cn'
import type { ButtonProps } from './types'
import { buttonVariants } from './helpers'
import { Icon } from '../icon'
import { LoaderCircleIcon } from 'lucide-react'

/**
 * Button component following browser-standard button behavior
 * Supports 4 variants: primary, ghost, outline, and destructive
 * Provides 3 sizes: sm, md (default), lg
 */
export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  function Button(
    {
      className,
      variant = 'primary',
      size = 'md',
      disabled = false,
      type = 'button',
      busy,
      children,
      ...props
    },
    ref
  ) {
    // Map button sizes to icon sizes
    const iconSize = size === 'sm' ? 'sm' : size === 'lg' ? 'md' : 'sm'

    // Determine icon color based on variant
    const getIconColor = () => {
      if (variant === 'primary' || variant === 'destructive') {
        return 'white'
      }
      if (variant === 'outline' || variant === 'ghost') {
        return 'inherit'
      }
      return 'inherit'
    }

    return (
      <button
        ref={ref}
        type={type}
        className={cn(buttonVariants({ variant, size }), className)}
        disabled={disabled || busy}
        aria-disabled={disabled || busy}
        inert={disabled || busy ? true : undefined}
        aria-busy={busy ? true : undefined}
        {...props}
      >
        {busy && (
          <Icon
            as={LoaderCircleIcon}
            size={iconSize}
            color={getIconColor()}
            className='animate-spin'
          />
        )}
        {children}
      </button>
    )
  }
)
