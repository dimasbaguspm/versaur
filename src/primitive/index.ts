export * from './alert'
export * from './anchor'
export * from './attribute'
export * from './attribute-list'
export * from './avatar'
export * from './badge'
export * from './brand'
export * from './button'
export * from './button-float'
export * from './button-icon'
export * from './calculator'
export * from './card'
export * from './filter-chip'
export * from './icon'
export * from './no-results'
export * from './snackbar'
export * from './table'
export * from './text'
export * from './tile'
export * from './image'
