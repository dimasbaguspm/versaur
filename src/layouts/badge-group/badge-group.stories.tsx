import type { Meta, StoryObj } from '@storybook/react'
import { BadgeGroup } from './badge-group'
import { Badge } from '@/primitive/badge'

/**
 * BadgeGroup provides layout management for badges with control over positioning,
 * alignment, and fluid behavior. It's designed to wrap Badge components and manage
 * their positioning effectively.
 *
 * ## Features
 * - **Alignment control**: Position badges with start, center, end, between, around, or evenly
 * - **Orientation**: Horizontal or vertical layout
 * - **Fluid behavior**: Make badges expand to fill available space
 * - **Gap management**: Control spacing between badges with predefined sizes
 * - **Accessibility**: Proper ARIA roles for screen readers
 *
 * ## Usage
 * Wrap Badge components with BadgeGroup to control their layout and alignment
 */
const meta: Meta<typeof BadgeGroup> = {
  title: 'Layouts/BadgeGroup',
  component: BadgeGroup,
  parameters: {
    docs: {
      description: {
        component:
          'A flexible layout component for managing badge positioning, alignment, and fluid behavior.',
      },
    },
  },
  argTypes: {
    alignment: {
      control: { type: 'select' },
      options: ['start', 'center', 'end', 'between', 'around', 'evenly'],
      description: 'Alignment of buttons within the group',
    },
    orientation: {
      control: { type: 'select' },
      options: ['horizontal', 'vertical'],
      description: 'Layout direction of the button group',
    },
    gap: {
      control: { type: 'select' },
      options: ['xs', 'sm', 'md', 'lg', 'xl'],
      description: 'Size of the gap between buttons',
    },
    fluid: {
      control: { type: 'boolean' },
      description: 'Whether buttons should expand to fill available space',
    },
  },
  tags: ['autodocs'],
}

export default meta
type Story = StoryObj<typeof BadgeGroup>

/**
 * Default badge badge with horizontal layout and start alignment
 */
export const Default: Story = {
  render: args => (
    <BadgeGroup {...args}>
      <Badge color='primary'>Primary</Badge>
      <Badge color='secondary'>Secondary</Badge>
      <Badge color='tertiary'>Tertiary</Badge>
      <Badge color='ghost'>Ghost</Badge>
      <Badge color='neutral'>Neutral</Badge>
    </BadgeGroup>
  ),
}

/**
 * Button group with center alignment - useful for modal actions
 */
export const CenterAligned: Story = {
  render: args => (
    <BadgeGroup {...args} alignment='center'>
      <Badge color='primary'>Primary</Badge>
      <Badge color='secondary'>Secondary</Badge>
      <Badge color='tertiary'>Tertiary</Badge>
      <Badge color='ghost'>Ghost</Badge>
      <Badge color='neutral'>Neutral</Badge>
    </BadgeGroup>
  ),
}

/**
 * Button group with space between - common for navigation or form actions
 */
export const SpaceBetween: Story = {
  render: args => (
    <div style={{ width: '400px', border: '1px solid var(--color-primary)' }}>
      <BadgeGroup {...args} alignment='between'>
        <Badge color='primary'>Primary</Badge>
        <Badge color='secondary'>Secondary</Badge>
        <Badge color='tertiary'>Tertiary</Badge>
        <Badge color='ghost'>Ghost</Badge>
        <Badge color='neutral'>Neutral</Badge>
      </BadgeGroup>
    </div>
  ),
}

/**
 * Vertical button group - useful for sidebars or stacked actions
 */
export const Vertical: Story = {
  render: args => (
    <BadgeGroup {...args} orientation='vertical'>
      <Badge color='primary'>Primary</Badge>
      <Badge color='secondary'>Secondary</Badge>
      <Badge color='tertiary'>Tertiary</Badge>
      <Badge color='ghost'>Ghost</Badge>
      <Badge color='neutral'>Neutral</Badge>
    </BadgeGroup>
  ),
}

/**
 * Fluid buttons that expand to fill available space
 */
export const Fluid: Story = {
  render: args => (
    <div style={{ width: '400px', border: '1px solid var(--color-primary)' }}>
      <BadgeGroup {...args} fluid>
        <Badge color='primary'>Primary</Badge>
        <Badge color='secondary'>Secondary</Badge>
        <Badge color='tertiary'>Tertiary</Badge>
        <Badge color='ghost'>Ghost</Badge>
        <Badge color='neutral'>Neutral</Badge>
      </BadgeGroup>
    </div>
  ),
}

/**
 * Vertical fluid layout - perfect for mobile interfaces
 */
export const VerticalFluid: Story = {
  render: args => (
    <div style={{ width: '200px' }}>
      <BadgeGroup {...args} orientation='vertical' fluid>
        <Badge color='primary'>Primary</Badge>
        <Badge color='secondary'>Secondary</Badge>
        <Badge color='tertiary'>Tertiary</Badge>
        <Badge color='ghost'>Ghost</Badge>
        <Badge color='neutral'>Neutral</Badge>
      </BadgeGroup>
    </div>
  ),
}

/**
 * Different gap sizes to control spacing
 */
export const GapVariations: Story = {
  render: _args => (
    <div className='space-y-4'>
      <div>
        <h3 className='mb-2 text-sm font-medium'>Extra Small Gap</h3>
        <BadgeGroup gap='xs'>
          <Badge color='primary'>Primary</Badge>
          <Badge color='secondary'>Secondary</Badge>
          <Badge color='tertiary'>Tertiary</Badge>
          <Badge color='ghost'>Ghost</Badge>
          <Badge color='neutral'>Neutral</Badge>
        </BadgeGroup>
      </div>
      <div>
        <h3 className='mb-2 text-sm font-medium'>Small Gap</h3>
        <BadgeGroup gap='sm'>
          <Badge color='primary'>Primary</Badge>
          <Badge color='secondary'>Secondary</Badge>
          <Badge color='tertiary'>Tertiary</Badge>
          <Badge color='ghost'>Ghost</Badge>
          <Badge color='neutral'>Neutral</Badge>
        </BadgeGroup>
      </div>
      <div>
        <h3 className='mb-2 text-sm font-medium'>Medium Gap (Default)</h3>
        <BadgeGroup gap='md'>
          <Badge color='primary'>Primary</Badge>
          <Badge color='secondary'>Secondary</Badge>
          <Badge color='tertiary'>Tertiary</Badge>
          <Badge color='ghost'>Ghost</Badge>
          <Badge color='neutral'>Neutral</Badge>
        </BadgeGroup>
      </div>
      <div>
        <h3 className='mb-2 text-sm font-medium'>Large Gap</h3>
        <BadgeGroup gap='lg'>
          <Badge color='primary'>Primary</Badge>
          <Badge color='secondary'>Secondary</Badge>
          <Badge color='tertiary'>Tertiary</Badge>
          <Badge color='ghost'>Ghost</Badge>
          <Badge color='neutral'>Neutral</Badge>
        </BadgeGroup>
      </div>
      <div>
        <h3 className='mb-2 text-sm font-medium'>Extra Large Gap</h3>
        <BadgeGroup gap='xl'>
          <Badge color='primary'>Primary</Badge>
          <Badge color='secondary'>Secondary</Badge>
          <Badge color='tertiary'>Tertiary</Badge>
          <Badge color='ghost'>Ghost</Badge>
          <Badge color='neutral'>Neutral</Badge>
        </BadgeGroup>
      </div>
    </div>
  ),
}
