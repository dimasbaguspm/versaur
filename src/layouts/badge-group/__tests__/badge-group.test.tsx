import { describe, it, expect } from 'vitest'
import { render, screen } from '@testing-library/react'
import { composeStories } from '@storybook/react'
import * as stories from '../badge-group.stories'
import { BadgeGroup } from '../badge-group'
import { Button } from '@/primitive/button'

const composedStories = composeStories(stories)

describe('BadgeGroup', () => {
  it('should render HTML correctly', () => {
    const { asFragment } = render(
      <BadgeGroup>
        <Button variant='primary'>Save</Button>
        <Button variant='ghost'>Cancel</Button>
      </BadgeGroup>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render with default props', () => {
    render(
      <BadgeGroup data-testid='button-group'>
        <Button>Button 1</Button>
        <Button>Button 2</Button>
      </BadgeGroup>
    )

    const badgeGroup = screen.getByTestId('button-group')
    expect(badgeGroup).toBeInTheDocument()
    expect(badgeGroup).toHaveAttribute('role', 'group')
    expect(badgeGroup).toHaveClass('flex', 'flex-row', 'justify-start', 'gap-3')
  })

  it('should apply vertical orientation', () => {
    render(
      <BadgeGroup orientation='vertical' data-testid='button-group'>
        <Button>Button 1</Button>
        <Button>Button 2</Button>
      </BadgeGroup>
    )

    const badgeGroup = screen.getByTestId('button-group')
    expect(badgeGroup).toHaveClass('flex-col')
  })

  it('should apply different alignments', () => {
    const { rerender } = render(
      <BadgeGroup alignment='center' data-testid='button-group'>
        <Button>Button 1</Button>
      </BadgeGroup>
    )

    let badgeGroup = screen.getByTestId('button-group')
    expect(badgeGroup).toHaveClass('justify-center')

    rerender(
      <BadgeGroup alignment='end' data-testid='button-group'>
        <Button>Button 1</Button>
      </BadgeGroup>
    )

    badgeGroup = screen.getByTestId('button-group')
    expect(badgeGroup).toHaveClass('justify-end')

    rerender(
      <BadgeGroup alignment='between' data-testid='button-group'>
        <Button>Button 1</Button>
      </BadgeGroup>
    )

    badgeGroup = screen.getByTestId('button-group')
    expect(badgeGroup).toHaveClass('justify-between')
  })

  it('should apply fluid behavior', () => {
    render(
      <BadgeGroup fluid data-testid='button-group'>
        <Button>Button 1</Button>
        <Button>Button 2</Button>
      </BadgeGroup>
    )

    const badgeGroup = screen.getByTestId('button-group')
    expect(badgeGroup).toHaveClass('[&>*]:flex-1')
  })

  it('should apply different gap sizes', () => {
    const { rerender } = render(
      <BadgeGroup gap='xs' data-testid='button-group'>
        <Button>Button 1</Button>
      </BadgeGroup>
    )

    let badgeGroup = screen.getByTestId('button-group')
    expect(badgeGroup).toHaveClass('gap-1')

    rerender(
      <BadgeGroup gap='xl' data-testid='button-group'>
        <Button>Button 1</Button>
      </BadgeGroup>
    )

    badgeGroup = screen.getByTestId('button-group')
    expect(badgeGroup).toHaveClass('gap-6')
  })

  it('should forward ref correctly', () => {
    const ref = { current: null }
    render(
      <BadgeGroup ref={ref}>
        <Button>Button 1</Button>
      </BadgeGroup>
    )

    expect(ref.current).toBeInstanceOf(HTMLDivElement)
  })

  it('should accept custom className', () => {
    render(
      <BadgeGroup className='custom-class' data-testid='button-group'>
        <Button>Button 1</Button>
      </BadgeGroup>
    )

    const badgeGroup = screen.getByTestId('button-group')
    expect(badgeGroup).toHaveClass('custom-class')
  })

  it('should render children correctly', () => {
    render(
      <BadgeGroup>
        <Button>Save</Button>
        <Button>Cancel</Button>
        <Button>Reset</Button>
      </BadgeGroup>
    )

    expect(screen.getByRole('button', { name: 'Save' })).toBeInTheDocument()
    expect(screen.getByRole('button', { name: 'Cancel' })).toBeInTheDocument()
    expect(screen.getByRole('button', { name: 'Reset' })).toBeInTheDocument()
  })

  describe('Stories', () => {
    it('should render Default story', () => {
      const { container } = render(<composedStories.Default />)
      expect(container.firstChild).toBeInTheDocument()
    })

    it('should render CenterAligned story', () => {
      const { container } = render(<composedStories.CenterAligned />)
      expect(container.firstChild).toBeInTheDocument()
    })

    it('should render SpaceBetween story', () => {
      const { container } = render(<composedStories.SpaceBetween />)
      expect(container.firstChild).toBeInTheDocument()
    })

    it('should render Vertical story', () => {
      const { container } = render(<composedStories.Vertical />)
      expect(container.firstChild).toBeInTheDocument()
    })

    it('should render Fluid story', () => {
      const { container } = render(<composedStories.Fluid />)
      expect(container.firstChild).toBeInTheDocument()
    })

    it('should render VerticalFluid story', () => {
      const { container } = render(<composedStories.VerticalFluid />)
      expect(container.firstChild).toBeInTheDocument()
    })

    it('should render GapVariations story', () => {
      const { container } = render(<composedStories.GapVariations />)
      expect(container.firstChild).toBeInTheDocument()
    })
  })
})
