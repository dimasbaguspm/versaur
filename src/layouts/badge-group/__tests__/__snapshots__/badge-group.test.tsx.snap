// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BadgeGroup > should render HTML correctly 1`] = `
<DocumentFragment>
  <div
    class="flex flex-row justify-start gap-3"
    role="group"
  >
    <button
      aria-disabled="false"
      class="inline-flex items-center justify-center gap-2 rounded-md font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none cursor-pointer bg-primary text-white hover:bg-primary/90 focus-visible:ring-primary-light focus-visible:ring-offset-white shadow-sm hover:shadow-md h-9 px-4 text-sm min-w-[2.5rem]"
      type="button"
    >
      Save
    </button>
    <button
      aria-disabled="false"
      class="inline-flex items-center justify-center gap-2 rounded-md font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none cursor-pointer bg-white text-foreground hover:bg-ghost-soft focus-visible:ring-ghost-light focus-visible:ring-offset-white h-9 px-4 text-sm min-w-[2.5rem]"
      type="button"
    >
      Cancel
    </button>
  </div>
</DocumentFragment>
`;
